Implemented component flexberrry-sitemap-searchbar

dpavloff opened the pull request:
This component can be integrated in any ember app created with ember-flexberry. It allows sitemap search using recursive algorithm. 

Documentation for structure of sitemap object is WIP.

KollegOFF commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+    this.set('results', this.get('treeContent'));
+    const $sitemapSearch = document.getElementById('sitemap-search');
+
+    this.$('.sitemap-searchbar.ui.search').on('click', e => {
+      this.$('#results-list').show();
+      e.stopPropagation();
+    });
+
+    this.$('#results-list').on('click', e => {
+      e.stopPropagation();
+    });
+
+    this.$(document).on('click', e => {
+      this.$('#results-list').hide();
+      e.stopPropagation();
+    });
+
+    $sitemapSearch.addEventListener('input', (e) => {
```
Это нужно делать через экшены, и их привязку к элементам в шаблоне.

dpavloff replied:
Оставил только событие `$(document).on('click)`, т.к. не могу найти другого способа проверять клик не по компоненту.

KollegOFF commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+
+    $sitemapSearch.addEventListener('input', (e) => {
+      let query = $sitemapSearch.value.toLowerCase();
+
+      if ((query === null || query === 'undefined' || query === '')) {
```
Слишком много скобок, и проверок, может просто `if (query) { /* ... */ } else { /* ... */ }`?

dpavloff replied:
Да, никак не поспоришь.

KollegOFF commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+
+      if ((query === null || query === 'undefined' || query === '')) {
+        this.set('results', this.get('treeContent'));
+      } else if (e.timeStamp - this.lastKeyPress > 200) { // Recursive search will be initiated only if last keypress happened more than 200 ms ago (for performance reasons).
+        let regexQuery = new RegExp(`${query}`, 'gi');
+
+        this.set('results', this._searchTree(regexQuery, this.get('treeContent')));
+        this.set('lastKeyPress', e.timeStamp);
+      }
```
Это нужно делать через [`Ember.run.debounce`](https://api.emberjs.com/ember/2.4/namespaces/Ember.run/methods/debounce?anchor=debounce):
```js
Ember.run.debounce(this, this._method, 200);
```

dpavloff replied:
Ого, не знал даже о такой функции

KollegOFF commented on app/templates/components/flexberry-sitemap-searchbar.hbs:
```diff
@@ -0,0 +1,12 @@
+<div class="sitemap-searchbar ui search">
+  <input id="sitemap-search" type="text" placeholder="Найти раздел" />
+  <div id="results-list">
```
Не нужно использовать айдишники для элементов, для стилей нужно использовать классы, для поиска элементов в контексте компонента `this.$('selector')`.
Текст для `placeholder` нужно вынести в локализацию.

dpavloff replied:
Вынес placeholder, убрал айдишники.

KollegOFF commented on app/templates/components/flexberry-sitemap-searchbar.hbs:
```diff
@@ -0,0 +1,12 @@
+  <input id="sitemap-search" type="text" placeholder="Найти раздел" />
+  <div id="results-list">
+    {{#each results as |node|}}
+    {{#searchbar-node
+    class="match-card"
+    node=node
+    }}
+    {{/searchbar-node}}
```
Не нужно писать компонент в блочной форме, если она не используется.
Классы может через `classNames` описать?

dpavloff replied:
Сделано

KollegOFF commented on app/templates/components/searchbar-node.hbs:
```diff
@@ -0,0 +1,24 @@
+      {{/each}}
+    {{/flexberry-toggler}}
+{{else}}
+ {{#link-to node.link title=node.caption.string}}
+    {{#if node.caption.string}}
+      {{node.caption.string}}
+    {{else}}
+      {{node.title.string}}
+    {{/if}}
+  {{/link-to}}
```
Нужно использовать `t` хелпер для вывода локализованных значений.
Для `title` у ссылки `node.title`, для содержимого `node.caption`.

KollegOFF commented on app/templates/sitemap.hbs:
```diff
@@ -1,3 +1,5 @@
+{{flexberry-sitemap-searchbar
+treeContent=model.nodes}}
```
Не нужно добавлять компонент сразу в этот шаблон, лучше добавить его где то здесь:
https://github.com/Flexberry/ember-flexberry/blob/d05d50a6226f580e04db8d2598d2d820db392113/tests/dummy/app/templates/application.hbs#L13-L19

dpavloff replied:
Перенёс в application.hbs.

KollegOFF reviewed (dismissed):
Нужны тесты.

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+    @property treeContent
+    @type Array
+  */
+  treeContent: Ember.A(),
```
Правда, так будет лучше называться. Поменял на sitemap. Начальное значение установил как null.

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+    @property results
+    @type Array
+  */
+  results: Ember.A(),
```
Заменил на null, сделал приватным.

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+   @private
+   @method _searchTree
+ */
+  _searchTree(regexQuery, currentTree) {
```
Я выбрал регулярное выражение, чтобы в будущем было проще реализовать подсвечивание попадаемых строк. Мне кажется, что такое решение будет очень наглядно и полезно.

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+    let resultTree = [];
+
+    currentTree.forEach(element => {
+      if (this._elementMatchesRegex(regexQuery, element)) {
+        resultTree.push(element);
+      } else if (this._elementHasChildren(element)) {
```
Не совсем понял, зачем нужна копия `element`'a?

dpavloff replied:
Записываю копию `element` с помощью lodash.

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+        let resultChildren = this._searchTree(regexQuery, element.children);
+
+        if (resultChildren.length > 0) {
+          let newElement = JSON.parse(JSON.stringify(element));
+          Ember.set(newElement, 'children', resultChildren);
+          resultTree.push(newElement);
```
Отличная идея, поменял на способ с деструктуризацией

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+  /**
+    Initializes DOM-related component's logic.
+  */
+  didInsertElement() {
+    this.set('results', this.get('treeContent'));
```
Хм, мне кажется что на первоначальном клике пользователю должна показываться полная менюшка, чтобы он был убежден в том, что поиск работает и можно вводить строку. Думаешь, что это так себе идея?

dpavloff replied:
Убрал.

dpavloff commented on addon/components/flexberry-sitemap-searchbar.js:
```diff
@@ -0,0 +1,109 @@
+   @private
+   @method _elementHasChildren
+ */
+  _elementHasChildren(element) {
+    return (typeof element === 'object') && (typeof element.children !== 'undefined') && (element.children !== null) && (element.children.length > 0);
+  },
```
Поменял последние три условия на `isEmpty(element.children)`

dpavloff committed:
Implemented flexberry-sitemap-searchbar

This component can be integrated in any ember app created with ember-flexberry. It allows sitemap search using recursive algorithm.

Documentation for structure of sitemap object is WIP.

TeterinaSvetlana committed:
Merge branch 'develop' into feature-196267

TeterinaSvetlana committed:
updated, test added

TeterinaSvetlana committed:
test fixed

TeterinaSvetlana committed:
fix file name

turbcool commented:
Документация https://github.com/Flexberry/flexberry.github.io/pull/121

pepelyaeva commented:
Нужно исправить под 3 версию

TeterinaSvetlana committed:
Merge branch 'develop' into feature-196267

TeterinaSvetlana committed:
fixes

TeterinaSvetlana committed:
fix error

TeterinaSvetlana committed:
remove extra template

TeterinaSvetlana committed:
add test into eslint exceptions

TeterinaSvetlana committed:
fix file name in eslint exceptions

TeterinaSvetlana committed:
fix lint

TeterinaSvetlana committed:
Fix destroyed object error

TeterinaSvetlana committed:
Add toggle searchbar

TeterinaSvetlana committed:
fix lint

pepelyaeva commented on app/templates/components/flexberry-sitemap-searchbar.hbs:
```diff
@@ -0,0 +1,36 @@
+<div class="sitemap-searchbar ui search">
```
Это же в classNames вынесено уже. Обертку можно удалить

pepelyaeva commented on app/templates/components/flexberry-sitemap-searchbar.hbs:
```diff
@@ -0,0 +1,36 @@
+<div class="sitemap-searchbar ui search">
+  {{input
+  class="sitemap-search-input"
+  type="text"
+  input=(action "startSearch")
+  click=(action "toggleResultsList")
+  value=userQuery
+  placeholder=placeholder
```
форматирование

pepelyaeva commented on app/templates/components/flexberry-sitemap-searchbar.hbs:
```diff
@@ -0,0 +1,36 @@
+    <div class="sitemap-search-results-list">
+      {{#each _results as |node|}}
+        {{searchbar-node
+        node=node
+        classNames="match-card"
+      }}
```
форматирование

pepelyaeva commented on app/templates/components/flexberry-sitemap-searchbar.hbs:
```diff
@@ -0,0 +1,36 @@
+        <div class="ui message">
+          <div class="header">
+            {{t "components.flexberry-sitemap-searchbar.errMsgOne"}}
+            <span>
+              {{userQuery}}
+            </span>
```
Лучше это в одну строку записать

pepelyaeva commented on tests/dummy/app/templates/application.hbs:
```diff
@@ -12,6 +12,8 @@
       onHidden=(action "updateWidth")
     )
   }}
+    {{flexberry-sitemap-searchbar
+    sitemap=sitemap.nodes}}
```
Лучше в одну строку

pepelyaeva commented on addon/components/flexberry-sitemap-searchbar.js:
Надо добавить скрытие при потере фокуса, а то я перешла на страницу из поиска а поле так и осталось открытым
![image](https://github.com/Flexberry/ember-flexberry/assets/15091209/5e675d02-9d52-43cd-adf7-d3901b7c10fc)

pepelyaeva commented on addon/components/flexberry-sitemap-searchbar.js:
Еще такой сценарий:
Свернуть меню, нажать на иконку поиска, кликнуть на любом месте формы - скрывается список с результатами.
Кликнуть еще раз по иконке поиска - скрывается поле ввода но отображается список с результатами

pepelyaeva commented on .eslint-todo-errors.json:
```diff
@@ -1925,6 +1925,13 @@
     { "ruleId": "ember/no-test-this-render", "line": 12, "column": 3 },
     { "ruleId": "hbs/check-hbs-template-literals", "line": 12, "column": 15 }
   ],
+  "tests/integration/components/flexberry-sitemap-searchbar-test.js": [
+    { "ruleId": "ember/no-test-module-for", "line": 13, "column": 1 },
+    { "ruleId": "ember/no-test-this-render", "line": 605, "column": 3 },
+    { "ruleId": "hbs/check-hbs-template-literals", "line": 605, "column": 15 },
+    { "ruleId": "ember/no-test-and-then", "line": 620, "column": 3 },
+    { "ruleId": "ember/no-test-and-then", "line": 628, "column": 5 }
+  ],
```
А вот эти все замечания надо бы поправить

pepelyaeva commented on .template-lintrc.js:
```diff
@@ -257,6 +257,12 @@ module.exports = {
       ]
     },
+    {
+      "moduleId": "app/templates/components/flexberry-sitemap-searchbar",
+      "only": [
+        "no-invalid-interactive"
+      ]
+    },
```
и это поправить, а не писать в исключения

TeterinaSvetlana committed:
fix lint hbs

TeterinaSvetlana committed:
fix lint hbs and test

TeterinaSvetlana committed:
code review

TeterinaSvetlana committed:
add class for button

TeterinaSvetlana committed:
remove extra text

TeterinaSvetlana committed:
Merge branch 'develop' into feature-196267

TeterinaSvetlana committed:
fix lint
